### ISSUE TITLE
Alpine apk package commit inclusion

### ISSUE THREAD

**Issue by deitch** [anchore/syft#1250]
**What would you like to be added**:

Include the alpine apk commit.

**Why is this needed**:

Here is sample output in spdx-json format:

```json
  {
   "SPDXID": "SPDXRef-837491105c32ac6d",
   "name": "findmnt",
   "licenseConcluded": "GPL-3.0-or-later AND GPL-2.0-or-later AND GPL-2.0-only",
   "description": "findmnt tool from util-linux",
   "downloadLocation": "https://git.kernel.org/cgit/utils/util-linux/util-linux.git",
   "externalRefs": [
    {
     "referenceCategory": "SECURITY",
     "referenceLocator": "cpe:2.3:a:findmnt:findmnt:2.38-r1:*:*:*:*:*:*:*",
     "referenceType": "cpe23Type"
    },
    {
     "referenceCategory": "PACKAGE_MANAGER",
     "referenceLocator": "pkg:alpine/findmnt@2.38-r1?arch=x86_64&upstream=util-linux&distro=busybox-1.35.0",
     "referenceType": "purl"
    }
   ],
   "filesAnalyzed": false,
   "hasFiles": [
    "SPDXRef-2040648de1ae8492"
   ],
   "licenseDeclared": "GPL-3.0-or-later AND GPL-2.0-or-later AND GPL-2.0-only",
   "originator": "Person: Natanael Copa <ncopa@alpinelinux.org>",
   "sourceInfo": "acquired package info from APK DB: /lib/apk/db/installed",
   "versionInfo": "2.38-r1"
  },
```

Clearly, the image ran `apk add findmnt` and installed version 2.38-r1. The problem is lack of reliable reproducibility in alpine. Version 2.38-r1 may or may not change and may or may not have the same source. More to the point, there is no pointer to the source there or specific commits, which can create other issues.

Fortunately, `/lib/apk/db/installed`, which Syft already scans to get the package information, includes the Alpine packages source commit:

```
C:Q1Y9xLLSQXfInWgA3aCH0+eXQq4+0=
P:findmnt
V:2.38-r1
A:x86_64
S:30375
I:81920
T:findmnt tool from util-linux
U:https://git.kernel.org/cgit/utils/util-linux/util-linux.git
L:GPL-3.0-or-later AND GPL-2.0-or-later AND GPL-2.0-only AND
o:util-linux
m:Natanael Copa <ncopa@alpinelinux.org>
t:1650054943
c:3f28965a3ed74837435bd0d488463d10afe72f3f
D:so:libblkid.so.1 so:libc.musl-x86_64.so.1 so:libmount.so.1 so:libsmartcols.so.1
p:cmd:findmnt=2.38-r1
i:util-linux=2.38-r1
F:bin
R:findmnt
a:0:0:755
Z:Q11cBWCxoRYDZsMQseu+s2ktVI7No=
```

The commit used to build that alpine package is `3f28965a3ed74837435bd0d488463d10afe72f3f`.

Could we include that, and perhaps even the Alpine linux package git URL, under `externalRefs`? It is `https://git.alpinelinux.org/aports/tree/main/${PACKAGENAME}/APKBUILD?id=${COMMIT}` so in the above example `https://git.alpinelinux.org/aports/tree/main/findmnt/APKBUILD?id= 3f28965a3ed74837435bd0d488463d10afe72f3f `

(actually, in my example, findmnt is part of `util-linux`, so it really is `https://git.alpinelinux.org/aports/tree/main/util-linux/APKBUILD?id= 3f28965a3ed74837435bd0d488463d10afe72f3f `, but same idea)

**Comment by deitch** [comment]
Oh, it also would make Syft much more helpful in compliance with licenses that require source reference.

**Comment by deitch** [comment]
Any takers? I even am willing to submit a PR (if maintainers agree in principle to accept, once it is ready).

**Comment by kzantow** [comment]
Hi @deitch, it looks like Syft is capturing this information already, just not including it in SPDX.

Looking through the [External Refs documentation](https://spdx.github.io/spdx-spec/v2.3/external-repository-identifiers/), there doesn't seem to be a category/type which would accurately describe this data. It's possible the [Source Information field](https://spdx.github.io/spdx-spec/v2.3/package-information/#712-source-information-field) would be more appropriate. However, I see there's a URL of https://git.kernel.org/cgit/utils/util-linux/util-linux.git which seems like it would be the right thing to use, but it doesn't contain commit 3f28965a3ed74837435bd0d488463d10afe72f3f, which as you point out is only in the aports git repository that doesn't seem to be listed here.

I don't think we would like to shoehorn incorrect data into SPDX.

Part of the issue there, is we should be able to generate SPDX and subsequently parse it back to a subset of the Syft data model accurately enough that we could subsequently regenerate the SPDX and get the same file back. This information would be lost without having a definitive location for it. Syft and CycloneDX formats already should include this information, though.

But perhaps if there was a consistent way to generate a link to the aports repo, that would work -- we could probably use that to parse the `o` and the `c` when reading SPDX.

**Comment by deitch** [comment]
> However, I see there's a URL of https://git.kernel.org/cgit/utils/util-linux/util-linux.git which seems like it would be the right thing to use

There is a more fundamental question here. If I have software which:

1. Comes from repository A (aports)
2. Which is just packaging up the true source from repository B (git.kernel.org/cgit/utils/...)

Which should be the source? I would argue that we should include both, if we can. But if not, it should be A. This software is an opinionated version of B, which may also have had patches or build args added, all while working in A. The source is A, and if you want to know about B, go look in A and find out about it.

> I don't think we would like to shoehorn incorrect data into SPDX.

Agreed, although we do not want to miss it either.

> Syft and CycloneDX formats already should include this information

Given the "standardness" of SPDX (at least as far as things that go near LF and its related worlds), I would be surprised if there isn't some way to get it in there.

> But perhaps if there was a consistent way to generate a link to the aports repo, that would work -- we could probably use that to parse the o and the c when reading SPDX

I didn't fully understand this. Share an example?

**Comment by deitch** [comment]
> But perhaps if there was a consistent way to generate a link to the aports repo, that would work -- we could probably use that to parse the o and the c when reading SPDX

I have been thinking about this @kzantow , as well as my comments about the proximate repo (aports) and upstream repos.

Would a reasonable compromise position be, "we will include in SPDX the aports commit as it appears in the `installed` file, but it is up to a consumer to determine how to use that"? That gets Syft out of the business of dealing with "which repo is correct or incorrect", while providing enough information for consumers to accurately track it.

When you have the commit in the SBoM, you have everything someone _needs_  to access the source. You haven't included every single step on the chain, the same way that you don't include upstream libraries, or build tools, or whatever else. "We have given you information, you figure out how to go use it."

**Comment by deitch** [comment]
> but it could maybe be something like `uses commit 3f28965a3ed74837435bd0d488463d10afe72f3f from aports fork of https://git.kernel.org/cgit/utils/util-linux/util-linux.git`

My view is much more constrained. If you get `commit 3f28965a3ed74837435bd0d488463d10afe72f3f` then you are covered. The fact that it is `pkg:alpk` and now includes a commit `3f28965a3ed74837435bd0d488463d10afe72f3f` to my mind is enough. I don't care that it references upstream https://git.kernel.org/cgit/utils/util-linux/util-linux.git, because `pkg:alpk` + `3f28965a3ed74837435bd0d488463d10afe72f3f` is enough. Getting from that to an aports URL, and from there to upstream https://git.kernel.org/cgit/utils/util-linux/util-linux.git, etc. is the responsibility of a consumer.

Besides, this likely is not an aports _fork_, but a commit of a repo where it references https://git.kernel.org/cgit/utils/util-linux/util-linux.git.

The other thing I don't like us `uses ...` because it is hard to machine parse.

**Comment by kzantow** [comment]
I agree about the semi-textual thing here... I'd sorta like to include the source repo, if we're already going to add information to this field. What about something like:
```
commit: 3f28965a3ed74837435bd0d488463d10afe72f3f source-repo: https://git.kernel.org/cgit/utils/util-linux/util-linux.git
```
I think that would be pretty easy to parse the commit or source-repo, if someone wanted that info -- thoughts?

**Comment by deitch** [comment]
Yeah it would be. I think that the source-repo might actually be several, I am wary of getting into what makes up the source in an `APKBUILD`, but since: a- you are including it from the `installed` file anyways (just taking what they give) and b- it is _more information_ - then sure.

**Comment by kzantow** [comment]
Maybe we could call this `upstream-repo` instead of `source-repo`?

**Comment by deitch** [comment]
If you want to do both, by all means. There is a pattern for getting the source-repo from the aports package name and commit, but I view that as a consumer responsibility.

**Comment by deitch** [comment]
Checking up on how this is going?
